L2H cold start tag predictions

cooglejj opened the pull request.

Commit:
L2H model for cold-start tag predictions

Commit:
Removed extraneous tag file

Zerthick commented on build.gradle:
```diff
@@ -46,5 +46,6 @@ dependencies {
     compile group: 'org.apache.logging.log4j', name: 'log4j-api', version: '2.8.2'
     compile group: 'org.apache.logging.log4j', name: 'log4j-core', version: '2.8.2'
     compile group: 'org.jasypt', name: 'jasypt', version: '1.9.2'
+	compile files('libs/segan.jar')
```
Adding a dependency directly on a file is generally considered a bad practice. However, if it is absolutely necessary (i.e. there is no external source that serves up the jar) you must include this jar in your commit so that it can be bundled with the project

cooglejj replied:
This is true. Unfortunately, the only way to get segan.jar right now is to manually download the source code and build it with specific instructions, so I figured this is the most practical solution. I've left this as is for now, but am open to suggestions.

Zerthick replied:
Yep. As I said, if it's unavoidable that this is what we have to do, we just have to make sure that the jar is included in the repo, otherwise when others pull it down they won't be able to build the project.

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/controller/component/stat/tags/UserTagStatComponent.java:
```diff
@@ -18,16 +23,48 @@
         this.userStatMap = ServiceManager.getService(project, PersistProfileComponent.class).getUserStatMap();
     }
 
+	public boolean createInitialTagPredictions(String filePath) {
```
Having this logic here is a bit strange. Wouldn't it make more sense to have this be a part of the PersonalSearchModel object itself?  Also, you are returning false in the case of an exception, however you never make any checks on that to handle the case where it's not possible to generate the initial predictions.

cooglejj replied:
Good point. I've relocated that to PersonalSearchModel. As for the return value, here's my thought process: On a high level, if the L2H model fails to load for some reason, the most sensible thing to do is run as normal without the initial predictions. In other words, there's nothing special to do if the predictions fail and it returns false. I still have the return value so we could do something in theory, but at this point I'm unclear what that would be.

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/controller/component/stat/tags/UserTagStatComponent.java:
```diff
@@ -18,16 +23,48 @@
 
+	public boolean createInitialTagPredictions(String filePath) {
+		try {
+			l2hPredictor.computeL2HPredictions(filePath, "./temp_predictions_file.txt", "./temp_text_data_file.txt");
```
Writing the predictions to a temp file to then be immediately read back in seems a bit strange. Where are theses files being written to on the user's machine? We don't want to be creating new files in their projects if we don't have to.  Would it be possible to return some sort of in-memory collection instead?

cooglejj replied:
Yes, this is the design aspect I was most unhappy about. Segan only allows predictions to move in or out of local files and provides no methods or parameters to put them anywhere else. Ideally, they would go in a collection, but there's no easy way to do that. In theory, it's possible to allow this by duplicating a number of segan's classes and adding custom code paths that work with collections instead of files. However, it's unclear how difficult that would be given how segan's design is based entirely around local file I/O. I could look into it if you feel the benefits would outweigh the time spent, code bloat, and additional code/design complexity.

damevski replied:
@cooglejj can you do some research into using pipes (or FIFOs) instead of actual files? this way you won't have to manage creating and deleting them, plus it should have better performance. my quick google search indicates that these aren't available for Java, but there could be a library or something we can use. 

https://www.gnu.org/software/libc/manual/html_node/Pipes-and-FIFOs.html

Zerthick replied:
I agree, we should see if there's a way to get around this issue. Writing to a temp file might end up being the solution we go with but it's worth some thought I think.  If we do end up writing to a temp file, we need to make sure it's placed in the appropriate directory along with the other plugin files like the user profile.

cooglejj replied:
No problem. I looked into what segan did and created a custom version of their code that does everything in memory, which of course means no more temp files. It does add some extra code, but not as much as I originally thought.

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/controller/component/stat/tags/UserTagStatComponent.java:
```diff
@@ -18,16 +23,48 @@
+			int countAmount = 5;
+			List<String> tagNames = new ArrayList<>();
+			List<Double> tagProbabilities = new ArrayList<>();
+			l2hPredictor.computeMostLikelyTags("./temp_predictions_file.txt", tagNames, tagProbabilities, maxTags);
```
Passing in collections to then be modified by a method is considered a bad practice, unless you are in the context of a consumer, which we are not here. Consider instead returning an object from this method that instead contains the desired collections to be utilized.

cooglejj replied:
Thanks for pointing that out, I was unaware that was a bad practice. I've created a custom class for tag predictions and have computeMostLikelyTags return a collection of that class instead.

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/controller/component/stat/tags/UserTagStatComponent.java:
```diff
@@ -18,16 +23,48 @@
+			List<Double> tagProbabilities = new ArrayList<>();
+			l2hPredictor.computeMostLikelyTags("./temp_predictions_file.txt", tagNames, tagProbabilities, maxTags);
+
+			List<String> tagToIncrement = new ArrayList<>();
```
You are storing the tags to increment here but aren't actually using them?

cooglejj replied:
They are being used a few lines below. tagToIncrement is passed to increaseTags.

Zerthick replied:
Ah I see, is it necessary to clear it then? That implies its state is persisting beyond the scope of the method which shouldn't really be the case.

cooglejj replied:
Yes. The logic is the increaseTags method will increase a collection of tags by a fixed amount. The createInitialTagPredictions method wants to increase a collection of tags, but by a different amount for each tag. So for each tag to increase, createInitialTagPredictions puts that tag in a List of just that one tag. Then it can use increaseTags with exactly the right amount for that one tag. The List is then cleared so it can be reused to do the same thing for the next tag (and only that next tag). That process continues until all tags have been increased by the correct amount.

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/controller/component/stat/tags/UserTagStatComponent.java:
```diff
@@ -18,16 +23,48 @@
     @Override
     public void projectOpened() {
         searchModel = new PersonalSearchModel(tagStatComponent, userStatMap);
+		try {
```
Since you are explicitly setting the predictor to null, it might be a good idea to do a null check before accessing it.  That way everything doesn't blow up with a NPE which has to be caught if something goes wrong.

cooglejj replied:
Good point. I've added a null check at the beginning.

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/model/L2H/CustomL2H.java:
```diff
@@ -0,0 +1,2231 @@
+package io.github.vcuswimlab.stackintheflow.model.L2H;
```
Does this file need to be a part of the plugin itself? (i.e. is it necessary for it to be bundled in the jar that we ship out to users?)  If it is only for use in a per-processing step to generate the model and it not utilized afterwards it should **not** be included in the main project repo.  We should instead create a sub-project for it.

cooglejj replied:
Yes, L2HPredictor uses the CustomL2H class to make it's predictions. It is not a pre-processing step. No pre-processing code is included in the pull request.

Zerthick replied:
You have a `public static void main` in here though. I doubt you are calling this class as a command line app while the plugin is running?  If possible you should create an output class that only contains the bare essentials needed while the plugin is running and commit that. Then create a generator script that produces the files needed by that class.

cooglejj replied:
That's true, the plugin doesn't need that particular part of the CustomL2H class. I've removed the main method and everything that was only used by the main method. It wasn't a drastic reduction, but less code is still less code.

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/model/L2H/L2HPredictor.java:
```diff
@@ -0,0 +1,182 @@
+        testSampler.setTestConfigurations(burnIn, modelIterations, sampleLag);
+	}
+
+	public void computeL2HPredictions(String fileToPredict, String outputFileName, String tempFileName) throws Exception {
```
I feel like instead of returning void these methods (computeL2HPredictions and computeMostLikelyTags) should be returning collections of some kind.

cooglejj replied:
I agree. computeMostLikelyTags is now returning a collection. computeL2HPredictions is still void for now because returning a collection there requires solving the 'writing predictions to temp file and reading back issue' that we discussed earlier.

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/model/L2H/L2HPredictor.java:
```diff
@@ -0,0 +1,182 @@
+        }
+	}
+
+	public void computeMostLikelyTags(String predictionsFileName, List<String> tagNamesOut, List<Double> tagProbabilitiesOut, int topN) throws Exception {
```
See above, I think it would be better to return an object containing the  tagNamesOut and tagProbabilitiesOut collections, rather than passing them in to be modified.

Zerthick replied:
Also, would it make more sense to return a sorted map of the tag name to the tag probability, that way they can easily be matched to one another?

cooglejj replied:
Agreed, computeL2HPredictions now returns a collection of a TagPrediction class that solves these issues.

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/model/L2H/L2HPredictor.java:
```diff
@@ -0,0 +1,182 @@
+package io.github.vcuswimlab.stackintheflow.model.L2H;
```
Check how you are doing file io here. There are better ways to get files out of the resources folder other than by using a buffered reader.  Take a look at the TagStatComponent for example, you should really be doing something like `this.getClass().getResourceLoader().getResource()` See https://stackoverflow.com/questions/3861989/preferred-way-of-loading-resources-in-java

cooglejj replied:
Thanks for pointing that out, I hadn't considered that could solve issues with using the jar over a network. I'm now getting resource streams like that. However, I'm accessing said resource streams through an InputStreamReader that is passed to a BufferedReader purely in interest of making minimal changes to the code. Is that okay? Or are there further issues with BufferedReader that I should be weary of?

Zerthick replied:
Buffered reader should be ok, when you are constructing the path of the file to read in it's also considered bad practice to concatenate strings, because different operating systems could use different path separators for example. Take a look at https://docs.oracle.com/javase/7/docs/api/java/nio/file/Paths.html#get(java.lang.String,%20java.lang.String...)

cooglejj replied:
Thanks for pointing that out as well, I hadn't thought about the code paths. I've switched file path logic to use the Path class where possible.

Zerthick reviewed (changes requested):
Do we need to include all of these files are part of the main plugin jar?  We don't want to bloat the download size with unnecessary files. We should only include files in the main repo that are necessary when the plugin is running on the user's machine. Not any model building scripts or required data for model building. We should create separate projects for that.

Commit:
Some revisions

cooglejj commented:
Thanks for the detailed feedback! I've responded to everything you mentioned and made changes accordingly. Please let me know if there's anything more I could do.

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/controller/component/stat/tags/UserTagStatComponent.java:
```diff
@@ -3,11 +3,14 @@
 import org.jetbrains.annotations.NotNull;
 
 import java.util.Map;
+import java.util.List;
```
Are these imports necessary?  Intellij has options to reformat and remove unnecessary imports on commit, you may want to turn those on if you haven't already.

cooglejj replied:
They are not, I simply forgot to remove them after I refactored. The unnecessary inputs in that class should be gone now.

Zerthick reviewed (commented):
Main thing to think about is how to get around writing to temp files. I'm also concerned that the size of the plugin jar might become bloated with all that additional files we will have to include. We should see if there is a way to trim down.

Commit:
Fixed temp files

cooglejj commented:
Thanks again for the detailed feedback. I've worked around needing temp files and resolved the other points as appropriate. If there's more I could do, please let me know.

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/model/L2H/L2HPredictor.java:
```diff
@@ -164,4 +197,13 @@ public void computeL2HPredictions(String fileToPredict, String outputFileName, S
 		return result;
 	}
+
+	public static void main(String[] args) throws Exception {
```
So I'm still confused why there needs to be a public static void main method anywhere in this project. Are you calling this while the plugin is running in the ide?

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/model/L2H/CustomL2H.java:
```diff
@@ -0,0 +1,2062 @@
+     * @param words Document words
+     * @param labels Document labels
+     */
+    public void train(ArrayList<Integer> docIndices, int[][] words, int[][] labels) {
```
So I don't think you are training a new model while the plugin is running? Should this logic be bundled with with plugin at all?

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/model/L2H/L2HPredictor.java:
```diff
@@ -0,0 +1,209 @@
+			builder.append(e.getKey() + ":" + e.getValue() + " ");
+		}
+
+		int[][] newWords = new TextDataLoader(new ByteArrayInputStream(builder.toString().getBytes())).getWords();
```
I'm confused as to why you are packing a map into a string just so you can then unpack it into a 2d array, there are much more efficient ways of doing this, for example https://stackoverflow.com/questions/2265266/convert-hash-map-to-2d-array

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/model/L2H/L2HPredictor.java:
```diff
@@ -0,0 +1,209 @@
+		return result;
+	}
+
+	public static void main(String[] args) throws Exception {
```
Again, I'm still confused what place a `public static void main` method has anywhere in the plugin's source. Are we calling part of this plugin as a command line app at runtime?

Zerthick commented on src/main/java/io/github/vcuswimlab/stackintheflow/model/L2H/CustomL2H.java:
```diff
@@ -0,0 +1,2062 @@
+		}
+    }
+
+	private static String outputSingleModelClassificationsToString(double[][] predictions) {
```
Serializing a 2d array to a string to be later deserialized seems a bit inefficient. Could you not just return the 2d array directly and skip the string serialization?

Zerthick commented:
We should only be including functionality that the plugin actually needs while running in the ide, anything that does not need to be executed on the user's machine should be bundled in separate scripts and put into somewhere like `/src/setup/java` instead of `/src/main/java`

Commit:
Minor changes

cooglejj commented:
Thanks again, I've removed the main method and done some other minor changes.

Something I think I should clarify: CustomL2H.java and the code in the TextDataLoader class (private class in L2HPredictor.java) are both directly copied from segan.jar. In the case of CustomL2H.java, it is a copy of segan's L2H.java, but with some code changes that work around some issues in the original L2H.java (also, the outputSingleModelClassificationToString method is copied with minor modifications from elsewhere in segan.jar). The code in TextDataLoader is copied from one of segan's text loading classes, but modified so it reads from an InputStream instead of a File (solving the temporary file issue we talked about).

With that in mind, I think some of these issues will make more sense. I'm using the InputStreams because it allows me to directly use segan's text loading code (in TextDataLoader) without writing a custom, specialized version for this specific use case. Given that, while I agree it is not the most efficient solution, I'm not convinced the relatively small loss in efficiency by using streams justifies creating a custom version tailored for this case. Notably, the current version works as-is if we need to work with predictions in files in the future, or even predictions from other sources such as an external website. Creating a specialized version would either remove that possibility entirely or result in further code duplication and complexity.

I think this code origin also explains why CustomL2H has methods that seemingly have nothing to do with actually running the model: I'm not trying to unnecessarily include code for building the model in the plugin. Rather, I'm taking an existing, very general class for L2H models, and trying to make the minimal necessary modifications for it to make correct predictions for the plugin. I've removed the train method, but I don't think there's much further refactoring that would add notable benefit. Given the copied nature of this class, the only reason I see for further refactoring is to reduce bloat from the jar. However, at this point the compiled class is only adding 51kb to the final jar. If bloat is a concern, I believe there are likely to be bigger and easier wins than stripping this class.

Of course, I could be overlooking something in these thoughts. Please let me know if you think there's more to it than what I've considered.

Zerthick commented:
Ah, this makes more sense then. Is there are reason you went with copying and modifying the code in the segan jar instead of simply extending the existing segan classes and overriding the methods that you need to modify?  I feel like that would make more sense from a sustainability point of view.

cooglejj commented:
Yes, I originally tried extending segan classes (You can see a minor remanent of that in the old version of TextDataLoader, which used to extend from a segan class). In the case of TextDataLoader, I needed to change the original method to use streams instead of files, so there's little that could be done in that case.

In the case of CustomL2H, the methods I needed to modify needed to be copied to begin with (or else the necessary modifications couldn't be performed), but they also directly or indirectly called nearly every private method in the original L2H class, meaning those methods would also need to be copied. Ultimately, it meant very little total code could be saved by extending, and I also didn't like the bug potential where overridden methods would call the copied version of the private methods, and non-overriden methods would call the original, uncopied version of the private methods. Therefore, I chose to copy both because of simplicity and because it was likely that a newer version of segan would eventually make the same fix I was doing, at which point the CustomL2H could be removed and replaced with the regular L2H class anyways.

damevski commented:
@cooglejj segan's code hasn't been modified for a couple of years ago, so, i think, most likely there won't be a better version. I think one of these two approaches for how to include L2H in StackInTheFlow makes sense to me, and please feel free to push back or for @Zerthick to chip in: 1) we modify Segan's code (as I think you have) and create a new customized jar that we pack with StackInTheFlow -- we maintain the source for the changed segan in some other repo; or 2) we strip *only* the stuff we absolutely need from segan, making our own de facto implementation of L2H (while attributing the code to segan) and have no dependency on his jar. Otherwise, as it looks to me, keeping more of segan's code than we strictly need, as source, feels like we are including and plan on maintaining some zombie part of segan's code.

Zerthick commented:
I agree with @damevski I think either option would be acceptable, though I'm a fan of option 2 since it means less overall bloat.

cooglejj commented:
I've investigated both approaches to some degree and come to some conclusions:

Approach 2 is undesirable for two reasons: Firstly, my efforts so far suggest it would take numerous hours to completely implement. Secondly, the stripping process has potential to introduce new bugs nearly anywhere in the segan code which would make them difficult to find and resolve.

Approach 1 is plausible, but it doesn't make as much sense under scrutiny. Notably, segan's L2H class is never used within the segan codebase. In other words, we can get the same effect by simply deleting the L2H class from segan.jar and using the CustomL2H class in it's place. At that point, the "changed segan in some other repo" would simplify to just the CustomL2H class by itself. Which is fine, but that then raises the question does it really make sense to have an entire separate repo devoted essentially to maintaining a single class?

In light of this, I think we should clarify what problem(s), specifically, we're trying to solve with these changes. I'm sure we all have a general sense, but I believe getting more specific will help us decide the best approach to take from this point.

damevski commented:
If Approach 2 isn't feasible, Approach 1 will have to do... Sounds like that approach consists of a packaging segan.jar and CustomL2H.java with StackInTheFlow. I do agree that my suggestion of a separate repo is pretty heavy handed. GitHub submodules could also be an option here, but I don't personally see the advantage of using them. Maybe one of you guys is a stronger advocate and can articulate how they would help us in the long run. 

If we go with the simplest approach of packaging these 2 files with the tool, I would consider renaming CustomL2H to SeganL2HAdapter (or something similar) and placing both the class and the jar in a clearly marked place in the StackInTheFlow repo. This also assumes that SeganL2HAdapter would be simple and clean, and not include any unnecessary bloat.

Commit:
Removed need for custom L2H class

Commit:
Removed need for extraneous util classes

cooglejj commented:
CustomL2H is now packaged inside the segan jar, and extraneous util classes are removed. Unless there's a further issue I've overlooked, I think this should be ready to use.
